### ISSUE TITLE
Iceberg fails to return results when filtered on complex columns

### ISSUE THREAD

**Issue by prodeezy** [apache/incubator-iceberg#99]
# Sample JSON Data
```
scala> val json = spark.read.schema(schema).json("people_complex.json").show()

+----+-------+--------------------+
| age|   name|             friends|
+----+-------+--------------------+
|null|Michael|                null|
|  30|   Andy|[Josh -> 10, Bria...|
|  19| Justin|[Bharat -> 15, Gau...|
+----+-------+--------------------+
```

# Json Table Schema
```
scala> json.printSchema
root
 |-- age: integer (nullable = true)
 |-- name: string (nullable = true)
 |-- friends: map (nullable = true)
 |    |-- key: string
 |    |-- value: integer (valueContainsNull = true)

```

# Predicate on  Map type column works on Parquet format with Vanilla spark reader 

```
scala> json.write.format("parquet").mode("append").save("parquet-people-complex")
scala> val parqDf = spark.read.format("parquet").load("parquet-people-complex")
scala> parqDf.createOrReplaceTempView("people_parquet_complex")

scala> val sqlDf = spark.sql("select * from people_parquet_complex where friends['Josh'] = 10").show()
sqlDf: org.apache.spark.sql.DataFrame = [age: int, name: string ... 2 more fields]
+---+----+--------------------+----------+
|age|name|             friends|  location|
+---+----+--------------------+----------+
| 30|Andy|[Josh -> 10, Gaut...|[101, 100]|
+---+----+--------------------+----------+

```

# Sample data in Iceberg format
```
scala> json.write.format("iceberg").mode("append").save("iceberg-people-complex")
scala> val iceDf = spark.read.format("iceberg").load("iceberg-people-complex").show()

+----+-------+--------------------+
| age|   name|             friends|
+----+-------+--------------------+
|null|Michael|                null|
|  30|   Andy|[Josh -> 10, Bria...|
|  19| Justin|[Bharat -> 15, Xa...|
+----+-------+--------------------+
```

# Filter on Complex (map) column fails to return result [This is the issue]
```
scala> iceDf.createOrReplaceTempView("people_iceberg_complex")
scala> spark.sql("select * from people_iceberg_complex where friends['Josh'] = 10").show()
sqlDF: org.apache.spark.sql.DataFrame = [age: int, name: string ... 1 more field]
+---+----+-------+
|age|name|friends|
+---+----+-------+
+---+----+-------+
```

**Comment by prodeezy** [comment]
I'm still trying to find root cause but at the outset this behavior seems incorrect. Will loop back with more findings.  Iceberg should at least let client handle the filtering (and not fail) or better still push down the predicate. 

If folks feel this is indeed a valid issue, will work on a proposal to fix this.

**Comment by prodeezy** [comment]
# Vanilla Spark Parquet reader plan
== Physical Plan ==
*(1) Project [age#428, name#429, friends#430, location#431]
+- *(1) Filter (isnotnull(friends#430) && (friends#430[Josh] = 10))
   +- *(1) FileScan parquet [age#428,name#429,friends#430,location#431] Batched: false, Format: Parquet, Location: InMemoryFileIndex[file:/usr/local/spark/test/parquet-people-complex], PartitionFilters: [], PushedFilters: [IsNotNull(friends)], ReadSchema: struct<age:int,name:string,friends:map<string,int>,location:struct<lat:int,lon:int>>

#  Iceberg Plan
== Physical Plan ==
*(1) Project [age#33, name#34, friends#35]
+- *(1) Filter ((friends#35[Josh] = 10) && isnotnull(friends#35))
   +- *(1) ScanV2 iceberg[age#33, name#34, friends#35] (Filters: [isnotnull(friends#35)], Options: [path=iceberg-people-complex2,paths=[]])


# Couple of points :
1)  Complex predicate is not passed down to the Scan level in both plans. The complex predicate is termed "non-translateable" by DataSourceStrategy.translateFilter()  [1] when trying to convert Catalyst expression to data source filter. Ryan & Xabriel had a discussion earlier on this list about Spark not passing expressions to data source (in certain cases). This might be related to that. Maybe a path forward is to fix that translation in Spark so that Iceberg Filter conversion has a chance to handle complex type. Currently Iceberg Reader code is unaware of that filter.

2) Although both vanilla Spark and Iceberg handle complex type predicates post scan, this regression is caused by post scan filtering not returning results in the Iceberg case. I think post scan filtering is unable to handle Iceberg format. So if 1) is not the way forward then the alternative way is to fix this in the post scan filtering.


[1] - https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L450

**Comment by rdblue** [comment]
@prodeezy, can you wrap the fixed-width sections with 3 back-ticks so they display properly? It's hard to read this.

**Comment by prodeezy** [comment]
@rdblue Sorry bout that. Done.

**Comment by prodeezy** [comment]
Gist with the example : https://gist.github.com/prodeezy/072dfbc69774652640e36b9ad5f17c68

**Comment by prodeezy** [comment]
Put up a PR and tested that it fixes the issue : https://github.com/apache/incubator-iceberg/pull/110

**Comment by aokolnychyi** [comment]
Can we close this issue as the PR was merged?

**Comment by rdblue** [comment]
Yes, this was fixed by #110. Thank you @prodeezy!
